File: UnivrCore/Sources/UnivrCore/Network.swift
//
//  Network.swift
//  Univr Core
//
//  Created by Leonardo Rossi on 08/10/25.
//  Copyright (C) 2026 Leonardo Rossi
//  SPDX-License-Identifier: GPL-3.0-or-later
//

import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct NetworkCacheData: Codable, Sendable {
    public let years: [AcademicYear]
    public let courses: [String: [Corso]]
}

@MainActor
public final class NetworkCache: Sendable {
    public static let shared = NetworkCache()
    
    public var years: [AcademicYear] = []
    public var courses: [String: [Corso]] = [:]
    
    private init() {}
    
    public func toData() -> NetworkCacheData {
        return NetworkCacheData(years: self.years, courses: self.courses)
    }
    
    public func update(from data: NetworkCacheData) {
        self.years = data.years
        self.courses = data.courses
    }
}

enum NetworkError: Error {
    case badURL
    case badServerResponse(statusCode: Int)
    case emptyData
    case decodingError(Error)
    case offline
    case timeout
    case unknown(Error)
    
    var errorDescription: String? {
        switch self {
        case .offline: return "Il dispositivo è offline."
        case .timeout: return "La richiesta è scaduta (Timeout)."
        case .badURL: return "L'URL non è valido."
        case .badServerResponse(let code): return "Errore Server: \(code)."
        case .emptyData: return "Nessun dato ricevuto dal server."
        case .decodingError(let err): return "Errore di decodifica: \(err.localizedDescription)"
        case .unknown(let err): return "Errore sconosciuto: \(err.localizedDescription)"
        }
    }
}

public protocol NetworkServiceProtocol: Sendable {
    func getYears() async throws -> [AcademicYear]
    func getCourses(year: String) async throws -> [Corso]
    func fetchOrario(corso: String, anno: String, selyear: String) async throws -> [DailySchedule]
}

public struct NetworkService: NetworkServiceProtocol {
    private let session: URLSession
    let baseURL = "http://192.168.0.20:3001/api/v1"
    
    public init() {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30
        
        // Da rendere per dispositivo
        configuration.httpAdditionalHeaders = [
            "User-Agent": "CalendarForUniVR/\(Bundle.main.clearAppVersion) (Device, OS)",
            "Accept": "application/json",
            "Content-Type": "application/json"
        ]
        
        self.session = URLSession(configuration: configuration)
    }
    
    private func fetch<T: Decodable>(from endpoint: String) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(endpoint)") else { throw NetworkError.badURL }
        
        do {
            let (data, response) = try await self.session.data(from: url)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw NetworkError.unknown(URLError(.unknown))
            }
            
            guard (200...299).contains(httpResponse.statusCode) else {
                throw NetworkError.badServerResponse(statusCode: httpResponse.statusCode)
            }
            
            let decoder = JSONDecoder()
            let formatter = DateFormatter()
            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
            formatter.locale = Locale(identifier: "en_US_POSIX")
            formatter.timeZone = TimeZone(identifier: "Europe/Rome")
            decoder.dateDecodingStrategy = .formatted(formatter)
            
            return try decoder.decode(T.self, from: data)
            
        } catch let error as URLError {
            switch error.code {
            case .notConnectedToInternet: throw NetworkError.offline
            case .timedOut: throw NetworkError.timeout
            default: throw NetworkError.unknown(error)
            }
        } catch let error as DecodingError {
            print("Decode error for \(endpoint): \(error)")
            throw NetworkError.decodingError(error)
        } catch {
            throw NetworkError.unknown(error)
        }
    }
    
    // MARK: - Public Methods
    public func getYears() async throws -> [AcademicYear] {
        struct RootWrapper: Decodable { let years: [AcademicYear] }
        let wrapper: RootWrapper = try await fetch(from: "/years")
        return wrapper.years
    }
    
    public func getCourses(year: String) async throws -> [Corso] {
        struct RootWrapper: Decodable { let courses: [Corso] }
        let wrapper: RootWrapper = try await fetch(from: "/courses?year=\(year)")
        return wrapper.courses
    }
    
    public func fetchOrario(corso: String, anno: String, selyear: String) async throws -> [DailySchedule] {
        return try await fetch(from: "/schedule?course=\(corso)&academicYear=\(anno)&year=\(selyear)")
    }
}

File: Android/core/src/main/java/dev/leonardorossi/univrcore/Network.kt
//
//  Network.kt
//  UnivrCore
//
//  Translated from Network.swift
//  Copyright (C) 2026 Leonardo Rossi
//  SPDX-License-Identifier: GPL-3.0-or-later
//

package dev.leonardorossi.univrcore

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.FormBody
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.net.UnknownHostException
import java.time.LocalDate
import java.util.concurrent.TimeUnit

// ── Cache dati di rete (in-memory, condiviso) ──

@Serializable
data class NetworkCacheData(
    val years: List<Year> = emptyList(),
    val courses: Map<String, List<Corso>> = emptyMap()
)

object NetworkCache {
    var years: List<Year> = emptyList()
    var courses: MutableMap<String, List<Corso>> = mutableMapOf()

    fun toData(): NetworkCacheData = NetworkCacheData(
        years = years,
        courses = courses.toMap()
    )

    fun update(data: NetworkCacheData) {
        years = data.years
        courses = data.courses.toMutableMap()
    }
}

// ── Errori di rete ──

sealed class NetworkError(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class BadURL : NetworkError("URL is not valid")
    class BadServerResponse(val statusCode: Int) : NetworkError("Server Error: $statusCode.")
    class EmptyData : NetworkError("Empty data received from the server")
    class DataNotFound(val variable: String) : NetworkError("Impossible to find data for: $variable.")
    class DecodingError(cause: Throwable) : NetworkError("Decoding error: ${cause.localizedMessage}", cause)
    class Offline : NetworkError("Device is offline.")
}

// ── Protocol / Interface ──

interface NetworkServiceProtocol {
    suspend fun getYears(): List<Year>
    suspend fun getCourses(year: String): List<Corso>
    suspend fun fetchOrario(corso: String, anno: String, selyear: String): ResponseAPI
}

// ── Implementazione OkHttp ──

class NetworkService : NetworkServiceProtocol {

    private val json = Json { ignoreUnknownKeys = true }

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .addInterceptor { chain ->
            val original = chain.request()
            val newRequest = original.newBuilder()
                .header("User-Agent", "CalendarForUniVR/${AppVersion.clearAppVersion} (Android)")
                .header("Accept", "application/json, text/html, */*")
                .header("Accept-Language", "it-IT,it;q=0.9,en;q=0.8")
                .build()
            chain.proceed(newRequest)
        }
        .build()

    companion object {
        private val yearsRegex = Regex("""var\s+anni_accademici_ec\s+=\s+(.*?);""", RegexOption.DOT_MATCHES_ALL)
        private val coursesRegex = Regex("""var\s+elenco_corsi\s+=\s+(.*?);""", RegexOption.DOT_MATCHES_ALL)
    }

    private fun buildUrl(endpoint: String = "combo.php", queryItems: Map<String, String>): String? {
        val base = "https://logistica.univr.it/PortaleStudentiUnivr/$endpoint".toHttpUrlOrNull()
            ?: return null
        val builder = base.newBuilder()
        queryItems.forEach { (k, v) -> builder.addQueryParameter(k, v) }
        return builder.build().toString()
    }

    private suspend inline fun <reified T> fetchAndExtract(
        queryItems: Map<String, String>,
        regex: Regex,
        variableName: String
    ): T {
        return withContext(Dispatchers.IO) {
            try {
                val url = buildUrl(queryItems = queryItems) ?: throw NetworkError.BadURL()

                val request = Request.Builder().url(url).get().build()
                val response = client.newCall(request).execute()

                if (!response.isSuccessful) {
                    throw NetworkError.BadServerResponse(response.code)
                }

                val text = response.body?.string() ?: throw NetworkError.EmptyData()

                val match = regex.find(text) ?: throw NetworkError.DataNotFound(variableName)
                val jsonString = match.groupValues[1]

                json.decodeFromString<T>(jsonString)
            } catch (e: UnknownHostException) {
                throw NetworkError.Offline()
            } catch (e: IOException) {
                throw NetworkError.Offline()
            } catch (e: NetworkError) {
                throw e
            } catch (e: Exception) {
                println("Decode error: $e")
                throw NetworkError.DecodingError(e)
            }
        }
    }

    override suspend fun getYears(): List<Year> {
        val yearsDict: Map<String, Year> = fetchAndExtract(
            queryItems = mapOf("aa" to "1"),
            regex = yearsRegex,
            variableName = "anni_accademici_ec"
        )
        return yearsDict.values.sortedBy { it.valore }
    }

    override suspend fun getCourses(year: String): List<Corso> {
        return fetchAndExtract(
            queryItems = mapOf("aa" to year, "page" to "corsi"),
            regex = coursesRegex,
            variableName = "elenco_corsi"
        )
    }

    override suspend fun fetchOrario(corso: String, anno: String, selyear: String): ResponseAPI {
        return withContext(Dispatchers.IO) {
            try {
                val formBody = FormBody.Builder()
                    .add("view", "easycourse")
                    .add("include", "corso")
                    .add("anno", selyear)
                    .add("cdl", corso)
                    .add("anno2", anno)
                    .add("_lang", "it")
                    .add("date", LocalDate.now().formatUnivrStyle())
                    .add("all_events", "1")
                    .build()

                val request = Request.Builder()
                    .url("https://logistica.univr.it/PortaleStudentiUnivr/grid_call.php")
                    .post(formBody)
                    .header("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8")
                    .build()

                val response = client.newCall(request).execute()

                if (!response.isSuccessful) {
                    throw NetworkError.BadServerResponse(response.code)
                }

                val body = response.body?.string() ?: throw NetworkError.EmptyData()
                json.decodeFromString<ResponseAPI>(body)
            } catch (e: UnknownHostException) {
                throw NetworkError.Offline()
            } catch (e: IOException) {
                throw NetworkError.Offline()
            } catch (e: NetworkError) {
                throw e
            } catch (e: Exception) {
                println("Decode error: $e")
                throw NetworkError.DecodingError(e)
            }
        }
    }
}
